ENH: Add cyclic garbage collection to NumPy

seberg opened the pull request:
This PR adds cyclic garbage collection to all (or at least most) objects within NumPy that can in theory form such cycles. It builds, and replaces, replaces gh-8303. Some thought may still be warrented in:

1. Give a bit of thoughts that all of these objects actually require cyclic garbage collection.
2. Ensure that the performance hit is insignificant (I assume it is, but should maybe time it).
    * Full test suit run seems unaffected on my machine. So I will assume that the performance hit is moderate to non-existing for most cases. (Which does not mean that it cannot be real for some use cases I suppose)
3. Probably a few more tests should be added to ensure that the cycle resolution always works as advertised.

seberg commented on numpy/core/src/multiarray/refcount.c.src:
```diff
@@ -14,46 +14,113 @@
+    // #define ERR_RETURN PyErr_WriteUnraisable(Py_None); return 0
+#else
+    #define CALL_INNER_FUNCTION(data, descr) @inner_function@(data, descr)
+    // #define ERR_RETURN return -1;
```
```suggestion
```
(also above). This file is the main annoyance. While it got quite a bit shorter by using the `c.src` the macros, etc. also added quite a bit of complexity.

Plus, I think there may have been some actual bug fixes here for which there are no test, but I do not remember what the fixes were :).

mattip replied:
you could do a `git blame` to see where it changed

seberg replied:
Oh, that was just me remembering to fix it up, it was my own change.

seberg commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -534,6 +545,8 @@ array_dealloc(PyArrayObject *self)
     Py_DECREF(fa->descr);
     Py_TYPE(self)->tp_free((PyObject *)self);
+
+    Py_TRASHCAN_SAFE_END(self)
```
Hmmm, I am not sure whether the other objects may require something similar. OTOH: only arrays can contain arrays. All other objects should always contain something which has the `TRASHCAN_SAFE` mechanism at some point I guess.

mattip commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1770,23 +1783,72 @@ array_iter(PyArrayObject *arr)
+     * or if it doesn't have any objects
+     */
+    if (!descr) {
+        return 0;
```
According to cov, the `!descr` path is never hit. Is it reasonable to remove the check?

seberg replied:
Changed it `PyArray_DATA(self) == NULL` for now, the issue is that allocation adds the object to the GC tracker too early on in a sense. It probably cannot happen that the traverse ever kicks in, since we are busy within C if it fails at this point. But that is what this seems to have been for (the change was by @J-Sand, maybe he has a quick note on it).

It is likely that this is just a band-aid solution, and the correct thing is to ensure that tracking is only enabled after everything is initialized. This might be a major rough edge :/ thought this was further along.

mattip commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -1290,6 +1301,9 @@ multiiter_new_impl(int n_args, PyObject **args)
             for (j = 0; j < mit->numiter; ++j) {
                 arr = (PyObject *)mit->iters[j]->ao;
+                if (PyDataType_REFCHK(PyArray_DESCR((PyArrayObject *)arr))) {
+                    needs_gc_tracking = NPY_TRUE;
```
This line is never hit.

mattip commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -1302,6 +1316,9 @@ multiiter_new_impl(int n_args, PyObject **args)
                 goto fail;
             }
+            if (PyDataType_REFCHK(PyArray_DESCR((PyArrayObject *)arr))) {
+                needs_gc_tracking = NPY_TRUE;
```
... but this one is

mattip commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -1869,14 +1911,16 @@ PyArray_NeighborhoodIterNew(PyArrayIterObject *x, const npy_intp *bounds,
-    Py_DECREF(ret->_internal_iter);
-    array_iter_base_dealloc((PyArrayIterObject*)ret);
-    PyArray_free((PyArrayObject*)ret);
+    Py_DECREF(ret);
```
This is never hit

mattip commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -1886,7 +1930,20 @@ static void neighiter_dealloc(PyArrayNeighborhoodIterObject* iter)
+        }
+    }
+    Py_VISIT(iter->_internal_iter);
+    return 0;
```
`neighiter_traverse` is never called

seberg replied:
I am considering to just undo the beighborhood iterator. It is an unused feature (within NumPy), which is meant to be used for C-side iteration of input arrays. (I bet scipy will be the only user).

mattip commented on numpy/core/src/multiarray/refcount.c.src:
```diff
@@ -14,46 +14,113 @@
+    NPY_COPY_PYOBJECT_PTR(data, &none);  \
+    Py_INCREF(none);                     \
+    Py_XDECREF(temp);                    \
+    } while (0)
```
Is this used? I couldn;t find where

mattip commented on numpy/core/src/multiarray/refcount.c.src:
```diff
@@ -14,46 +14,113 @@
+                     visitproc visit, void *arg)
+#else
+    @inner_function@(char *data, PyArray_Descr *descr)
+#endif
```
This is really hard to reason about. Maybe it would be better to expand some of these templates at the cost of some code repitition.

seberg replied:
changed slightly, not sure it is much better with inline `#defines` for the arguments though.

mattip commented on numpy/core/src/multiarray/refcount.c.src:
```diff
@@ -234,60 +180,87 @@ PyArray_XDECREF(PyArrayObject *mp)
-                NPY_COPY_PYOBJECT_PTR(&temp, data);
-                Py_XDECREF(temp);
+            for( i = 0; i < n; i++, data++) {
+                HANDLE_ITEM_AT_UNALIGNED(data);
```
I am pretty sure this is never hit

mattip commented on numpy/core/src/multiarray/refcount.c.src:
```diff
@@ -234,60 +180,87 @@ PyArray_XDECREF(PyArrayObject *mp)
+        while (it.index < it.size) {
+            data = (PyObject **)it.dataptr;
+            HANDLE_ITEM_AT_UNALIGNED(data);
             PyArray_ITER_NEXT(&it);
```
I don't think this is ever hit

seberg replied:
Probably not, it is pretty rare to create unaligned object arrays. Same goes for object arrays actually holding NULL instead of None. Although, I think the first one at least may be simple enough with a structured dtype, which would indeed be a good test.

seberg replied:
nvm, made a test for this, I think.

mattip commented on numpy/core/src/multiarray/refcount.c.src:
```diff
@@ -350,3 +323,38 @@ _fillobject(char *optr, PyObject *obj, PyArray_Descr *dtype)
+        n = PyArray_SIZE(arr);
+        if (obj == NULL) {
+            for (i = 0; i < n; i++) {
+                *optr++ = NULL;
```
The code path `obj == NULL` is never hit

seberg replied:
Yes, but it is exposed API, note that this function was not touched, the file was just modified so much it appears to be changed. Also it could just as well use `XINCREF` and remove the branching, I doubt it is worth to micro-optimize...

mattip reviewed (commented):
This seems somewhere between a cleanup and an enhancement. There are a lot of moving parts. I ran coverage locally and marked the new code paths that were not hit. I think `refcount.c.src` might need a little more thought: it is very dense and hard to reason about now.

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1770,23 +1783,72 @@ array_iter(PyArrayObject *arr)
+        /* if the array doesn't own its data, visiting fa->base is sufficient */
+        return 0;
+    }
+    /* NOTE: In the future, user dtypes could in principle be visited */
```
Shouldn't the dtype be visited anyway, since it has metadata that could refer back to the array? No need to implement the visiting in the dtype object yet, but I think we should call visit here to play it safe for when we fix that.

seberg replied:
It is pretty esoteric, but yes. We should visit the descriptor maybe. That would mean we cannot untrack arrays though, since any array has a dtype. (at least unless we also check that the dtype has no metadata). I think I meant the comment as visiting a user-dtypes items (which is much further off).

Not sure what to best do. Could just not do the untracking, or do it anyway, but visit. Either way, this would fix the vast majority of use-cases.

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1090,6 +1090,14 @@ PyArray_NewFromDescr_int(
+         * for object arrays. Subtypes may also define additional fields, so
+         * a subtype would have to handle this by itself.
+         */
+        PyObject_GC_UnTrack(fa);
```
Are you sure this is correct? What does cpython do about `dict` subclasses, for instance?

seberg replied:
To be honest, I have no thought to much about the untracking, it is the method I took from the original PR. But I am not sure how you think this can be incorrect?

seberg replied:
OK, I actually looked a bit at the Python code, and it seems to me that this is _not_ correct and I would have expected it to run into asserts on a debug python build (under the assumption that such an array is added to container that implements cyclic gc). I am a bit surprised I did not notice that (although without checking I am not 100% we have CI which runs a python debug build).

There would be solution to this: One can implement `tp_is_gc` which returns `False` after untracking/for arrays which are not added to the GC. (There is probably only a single use case of this: Python uses it to distinguish between HEAP and static types.)

Right now, I think this is arguably a bug, and chances are that having a few more arrays does not matter (there must be a huge number of other python containers around that need to do more traversing).

Still planning to try out the speed penalty for keeping a huge object array around, but even if it is noticable, I am not sure it matters too much as to whether or not we should do this.

eric-wieser replied:
Why is the object being tracked at this point already? From the docs:
```
PyObject_GC_Track` should be called once all the fields followed by the tp_traverse handler become valid, usually near the end of the constructor.
```
Could we just call `PyObject_GC_Track` here instead?

eric-wieser replied:
In fact, aren't we supposed to call `PyObject_GC_Track` ourselves anyway? I'm not seeing the bit of the docs that imply its called for us.

seberg replied:
The default `tp_alloc` calls it for you when `Py_TPFLAGS_HAVE_GC`. Strictly speaking: probably yes. But I also do not think it matters, since there we own the GIL the whole time here, and there is no call back into python before everything is finished up.

Anyway, if someone wants to review+merge this, I am happy to look at this, otherwise, for now we can also close until someone actually complains due to a memory leak. The new changes in python probably need a bit of cleaning, and the release notes may need to warn hypothetical hacks that subclass from C at least.

eric-wieser commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -1266,13 +1276,14 @@ static PyObject*
 
-    multi = PyArray_malloc(sizeof(PyArrayMultiIterObject));
+    multi = ((PyArrayMultiIterObject *)PyArrayMultiIter_Type.
+            tp_alloc(&PyArrayMultiIter_Type, 0));
```
Ideally you would pass through `subtype` from `arraymultiter_new`, and use it in place of `PyArrayMultiIter_Type` here

eric-wieser commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -1446,12 +1467,27 @@ arraymultiter_dealloc(PyArrayMultiIterObject *multi)
+arraymultiter_traverse(PyArrayMultiIterObject *multi, visitproc visit,
+        void *arg)
+{
+    int i;
```
nit: put this inside the `if`, or even in the `for` if not targeting a backport

eric-wieser commented on numpy/core/src/multiarray/nditer_pywrap.c:
```diff
@@ -1207,6 +1211,21 @@ npyiter_dealloc(NewNpyArrayIterObject *self)
+static int
+npyiter_traverse(NewNpyArrayIterObject *self, visitproc visit, void *arg)
+{
+    npy_intp iop, nop;
```
nit: push these down in scope

eric-wieser commented on numpy/core/src/multiarray/refcount.c.src:
```diff
@@ -14,46 +14,113 @@
+
+/* Use macros to achieve the different behaviour and function signatures */
+#if !@set_to_none@
+   #define HANDLE_ITEM_AT(data) @item_func@(*(data))
```
mismatched indent
```suggestion
    #define HANDLE_ITEM_AT(data) @item_func@(*(data))
```

seberg committed:
ENH: add cyclic GC support to ndarrays containing objects

This commit is based on work of James Sanders in:
https://github.com/numpy/numpy/pull/8303

Closes #6581.

Co-authored-by: James Sanders <J-Sand@users.noreply.github.com>

seberg commented:
Thought I might be able to unstick this a bit by limiting it to ndarray for now (the other part is preserved https://github.com/seberg/numpy/tree/cyclic-gc-iterators). But in any case the untracking is a pattern I have not found in cpython itself. However, it is completely fine to untrack (and even retrack), and it seems even fine to untrack multiple times.
Otherwise not much changed, but squashed (the untrack logic in `array_dealloc` is slightly adapted).

If this remains stuck, just as well...

seberg commented:
I suppose, the one real change to consider here is that might make sense to not bother doing this (possibly weird) untracking some arrays and just keep it tracked since the traverse will be very short.

mattip commented:
It is not clear how this might affect the entire interpreter performance if by chance you create a large object array in global namespace and never touch it again. Would it slow down the whole code?

mattip commented:
I am -1 on spending more effort on this now. I don't see a clear use case for adding this complexity, and it does not seem to be causing users much pain. We can revisit it if a user comes forward with a clear case where the perceived memory leak is meaningful.

seberg commented:
OK, fair enough.

seberg committed:
Fixup: Remove untracking and use new macros when available

This is probably right, but not really finished up. Just pushing
since it was almost finished, since if it seems like unless this
comes up again, we do not care enough about fixing this lingering
bug.

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
Perhaps add a `Python >= v3.8.0b1` comment here, so we know when to clean it up

seberg replied:
Yeah, should use Py_HEXVERSION I think, better to just compile time error...

seberg replied:
Comment is better, not sure I will go back to this, but stumbled over it and htought would quickly apply. If we anticipate array subclasses implemented in C, it might be necessary to add a compatibility release note for that.

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -534,8 +549,72 @@ array_dealloc(PyArrayObject *self)
+    /* Make sure we don't traverse the array before it is fully initialized */
+    if (PyArray_DATA(self) == NULL) {
+        return 0;
+    }
```
I think this should be an assert - if we manage to reach this, then we tracked the object too soon.

seberg committed:
Small fixups to address review comments

seberg commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -534,8 +549,72 @@ array_dealloc(PyArrayObject *self)
+     * this is guaranteed since the GC cannot run between creation and
+     * finishing initialization.
+     */
+    assert(PyArray_DATA(self) == NULL);
```
```suggestion
    assert(PyArray_DATA(self) != NULL);
```

seberg committed:
Update numpy/core/src/multiarray/arrayobject.c

charris commented:
@seberg Needs rebase. Do you want to keep it open.

seberg commented:
Well, if we can find someone to push it through, I am happy to rebase it.  It needs a little bit of thought, since I am not quite sure about those attempts to untrack arrays that don't (really) need it – i.e. the ones that don't contain objects.  Although, I know a thing or two more about it now, so maybe I can make sense of it now...

In general, I think it fixes a couple of impossible to find memory leaks even though they seem reported rarely...

seberg commented:
This would also be nice to revive.  I am not sure how much work it even is, the main thing to figure out (besides from trying to clean up the code a bit) is really how to best support subclasses.

seberg commented:
Wrong comment:  This should be revived IMO, it is a pretty big bug in most ways you can look at it.  I think it could be revived almost as-is (with some thoughts on whether we should even bother to selectively enable/disable tracking and how to do it correctly).

But, it may also be that sometime soon, and with the HPy project making progress the approach should change here anyway (i.e. the current decref/incref code paths being replaced).
